app/src/lib/git/spawn.ts

```diff
@@ -13,80 +14,83 @@ type ProcessOutput = {
+          if (
+            stdOutMaxLength &&
+            totalStdoutLength >= stdOutMaxLength &&
+            !killSignalSent
+          ) {
+            process.kill()
+            killSignalSent = true
+          }
+        })
 
-    const stdout = new Array<Buffer>()
-    let output: Buffer | undefined
-    process.stdout.on('data', chunk => {
-      stdout.push(chunk as Buffer)
-    })
+        const stderrChunks = new Array<Buffer>()
+        process.stderr.on('data', (chunk: Buffer) => {
+          stderrChunks.push(chunk)
+        })
 
-    const stderr = new Array<Buffer>()
-    let error: Buffer | undefined
-    process.stderr.on('data', chunk => {
-      stderr.push(chunk as Buffer)
-    })
+        process.on('error', err => {
+          // for unhandled errors raised by the process, let's surface this in the
+          // promise and make the caller handle it
+          reject(err)
+        })
 
-    function reportTimings() {
-      if (startTime) {
-        const rawTime = performance.now() - startTime
-        if (rawTime > 1000) {
-          const timeInSeconds = (rawTime / 1000).toFixed(3)
-          log.info(`Executing ${commandName} (took ${timeInSeconds}s)`)
-        }
-      }
-    }
-
-    process.stdout.once('close', () => {
-      // process.on('exit') may fire before stdout has closed, so this is a
-      // more accurate point in time to measure that the command has completed
-      // as we cannot proceed without the contents of the stdout stream
-      reportTimings()
-
-      output = Buffer.concat(stdout)
-      if (output && error) {
-        resolve({ output, error })
-      }
-    })
-
-    process.stderr.once('close', () => {
-      error = Buffer.concat(stderr)
+        process.on('close', (code, signal) => {
+          const stdout = Buffer.concat(
+            stdoutChunks,
+            stdOutMaxLength
+              ? Math.min(stdOutMaxLength, totalStdoutLength)
+              : totalStdoutLength
+          )
 
-      if (output && error) {
-        resolve({ output, error })
-      }
-    })
+          const stderr = Buffer.concat(stderrChunks)
 
-    process.on('error', err => {
-      // for unhandled errors raised by the process, let's surface this in the
-      // promise and make the caller handle it
-      reject(err)
-    })
+          // mimic the experience of GitProcess.exec for handling known codes when
+          // the process terminates
+          const exitCodes = successExitCodes || new Set([0])
 
-    process.on('exit', (code, signal) => {
-      // mimic the experience of GitProcess.exec for handling known codes when
-      // the process terminates
-      const exitCodes = successExitCodes || new Set([0])
-      if (!exitCodes.has(code)) {
-        reject(
-          new Error(
-            `Git returned an unexpected exit code '${code}' which should be handled by the caller.'`
-          )
-        )
-      }
-    })
-  })
+          if (exitCodes.has(code) || signal) {
```

Comment:
Footnote for anyone else who wasn't quite clear on this:

> If the process exited, `code` is the final exit code of the process, otherwise `null`. If the process terminated due to receipt of a signal, `signal` is the string name of the signal, otherwise `null`. **One of the two will always be non-null.**

[[source]](https://nodejs.org/dist/latest-v8.x/docs/api/child_process.html#child_process_event_exit)